File: idea-plugin/src/test/kotlin/dev/djarvur/c4drill/core/RenderParamsTest.kt
package dev.djarvur.c4drill.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Wire-parameter construction for c4drill/renderDiagram (mirrors renderParams.test.ts from #27). */
class RenderParamsTest {
    @Test
    fun `initial state is the c1 context view with model defaults`() {
        val params = buildRenderParams("file:///m.toml", initialViewState)

        assertEquals("file:///m.toml", params.textDocumentUri)
        assertEquals("", params.target)
        assertNull(params.allExpanded)
        assertNull(params.expanded) // omitted = model default
        assertNull(params.legend)
        assertEquals("svg", params.format)
    }

    @Test
    fun `all expanded overrides the target`() {
        val params = buildRenderParams(
            "file:///m.toml",
            PreviewViewState(target = "cloud.ui", allExpanded = true),
        )

        assertTrue(params.allExpanded!!)
        assertNull(params.target)
    }

    @Test
    fun `expanded text maps to an explicit expanded set`() {
        val params = buildRenderParams("u", PreviewViewState(expandedText = "cloud, db.io"))

        assertEquals(listOf("cloud", "db.io"), params.expanded)
    }

    @Test
    fun `blank expanded text keeps the model default`() {
        assertNull(buildRenderParams("u", initialViewState).expanded)
        assertNull(buildRenderParams("u", PreviewViewState(expandedText = "  ")).expanded)
        assertNull(buildRenderParams("u", PreviewViewState(expandedText = ",, ")).expanded)
    }

    @Test
    fun `collapse all sends the empty array override`() {
        val params = buildRenderParams("u", PreviewViewState(collapseAll = true, expandedText = "ignored"))

        assertEquals(emptyList<String>(), params.expanded)
    }

    @Test
    fun `legend overrides map to booleans`() {
        assertTrue(buildRenderParams("u", PreviewViewState(legend = LegendChoice.ON)).legend!!)
        assertFalse(buildRenderParams("u", PreviewViewState(legend = LegendChoice.OFF)).legend!!)
        assertNull(buildRenderParams("u", PreviewViewState(legend = LegendChoice.DEFAULT)).legend)
    }

    @Test
    fun `expanded set parsing splits on commas and whitespace`() {
        assertEquals(listOf("a", "b", "c"), parseExpandedSet("a, b\n\tc"))
        assertEquals(listOf("single"), parseExpandedSet("single"))
        assertNull(parseExpandedSet(""))
    }

    @Test
    fun `breadcrumb trail grows one entry per dotted segment`() {
        val trail = breadcrumbTrail(PreviewViewState(target = "cloud.ui"))

        assertEquals(listOf("C1", "cloud", "ui"), trail.map { it.label })
        assertEquals(listOf("", "cloud", "cloud.ui"), trail.map { it.target })
        assertFalse(trail[0].current)
        assertFalse(trail[1].current)
        assertTrue(trail[2].current)
    }

    @Test
    fun `breadcrumb trail for c1 and all-expanded`() {
        val c1 = breadcrumbTrail(initialViewState)

        assertEquals(1, c1.size)
        assertTrue(c1[0].current)

        val expanded = breadcrumbTrail(PreviewViewState(allExpanded = true))

        assertEquals(listOf("C1", "expanded"), expanded.map { it.label })
        assertTrue(expanded[1].current)
    }
}

File: extension/src/test/unit/renderParams.test.ts
import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import {
    breadcrumbTrail,
    buildRenderParams,
    initialViewState,
    parseExpandedSet,
} from '../../core/renderParams';

test('default view state renders the C1 context diagram', () => {
    const params = buildRenderParams('file:///w/model.toml', initialViewState);

    assert.deepEqual(params, {
        textDocument: { uri: 'file:///w/model.toml' },
        target: '',
        format: 'svg',
    });
});

test('a deep target is passed through for C3 rendering', () => {
    const params = buildRenderParams('file:///w/model.toml', {
        ...initialViewState,
        target: 'amazon.rds.pg',
    });

    assert.equal(params.target, 'amazon.rds.pg');
    assert.equal(params.allExpanded, undefined);
});

test('allExpanded overrides the target', () => {
    const params = buildRenderParams('file:///w/model.toml', {
        ...initialViewState,
        target: 'cloud',
        allExpanded: true,
    });

    assert.equal(params.allExpanded, true);
    assert.equal(params.target, undefined);
});

test('expanded text maps to the expanded array override', () => {
    const params = buildRenderParams('file:///w/model.toml', {
        ...initialViewState,
        expandedText: 'cloud, amazon.rds  next',
    });

    assert.deepEqual(params.expanded, ['cloud', 'amazon.rds', 'next']);
});

test('blank expanded text omits the field (model default)', () => {
    const params = buildRenderParams('file:///w/model.toml', {
        ...initialViewState,
        expandedText: '  ,  ',
    });

    assert.equal(params.expanded, undefined);
});

test('collapseAll sends the explicit empty-array override', () => {
    const params = buildRenderParams('file:///w/model.toml', {
        ...initialViewState,
        collapseAll: true,
        expandedText: 'cloud',
    });

    assert.deepEqual(params.expanded, []);
});

test('legend choices map to the boolean override or omission', () => {
    assert.equal(buildRenderParams('u', initialViewState).legend, undefined);
    assert.equal(buildRenderParams('u', { ...initialViewState, legend: 'on' }).legend, true);
    assert.equal(buildRenderParams('u', { ...initialViewState, legend: 'off' }).legend, false);
});

test('breadcrumbs chain from the C1 root', () => {
    const trail = breadcrumbTrail({ ...initialViewState, target: 'amazon.rds' });

    assert.deepEqual(
        trail.map((e) => [e.label, e.target, e.current]),
        [
            ['C1', '', false],
            ['amazon', 'amazon', false],
            ['rds', 'amazon.rds', true],
        ],
    );
});

test('C1 trail is a single current entry', () => {
    const trail = breadcrumbTrail(initialViewState);

    assert.equal(trail.length, 1);
    assert.equal(trail[0].current, true);
});

test('allExpanded mode appends the expanded marker', () => {
    const trail = breadcrumbTrail({ ...initialViewState, allExpanded: true });

    assert.deepEqual(
        trail.map((e) => e.label),
        ['C1', 'expanded'],
    );
    assert.equal(trail[1].current, true);
    assert.equal(trail[0].current, false);
});

test('parseExpandedSet is nil-for-blank and splits on mixed separators', () => {
    assert.equal(parseExpandedSet(''), undefined);
    assert.equal(parseExpandedSet('   '), undefined);
    assert.deepEqual(parseExpandedSet('a,b'), ['a', 'b']);
    assert.deepEqual(parseExpandedSet('a.b\n c ,\td'), ['a.b', 'c', 'd']);
});
